### ISSUE TITLE
[Rust] Implement three new Concept Exercises: intro-{fn, types, option}

### ISSUE THREAD

**Issue by workingjubilee** [exercism/rust#1059]
This issue breaks with the convention of a single issue per exercise because of the interrelatedness of these three exercises and the problem they intend to solve. They still need three patches, but it will likely benefit from a single issue for discussion as they rapidly build on each other and have a shared higher-level goal: overcoming some of the bigger newbie (and mentoring) Rust pain points, which are learning (or teaching) about functions, types, and unwrapping, as doing anything else with Rust depends on these.

Exercism is targeted at programmers of some experience but familiarity with the particular mix of C++ and OCaml idioms that Rust uses is not common, and even less so for students from further-out PLs like Python or Ruby. If Exercism is to successfully introduce Rust to others, it must quickly introduce the skillset that allows one to learn the rest of the language.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

These are intended to be model Rust exercises that show what is expected later on.
They are intended to teach the student enough to interface with Exercism.
They are intended to teach the student enough to interface with Rust's core APIs, so later exercises can assume an initial grasp of Rust usage and the student can go on learning how to interact with the rest of the Rust language.

The current intended divide is:
- `intro-fn`: teaches about type signatures and working with them, including some easy type casts. May use the pattern of [Rustlings]: provide "broken" code and request the reader to fix it after running `rustc` or `cargo check`, since main focus is on reading.
- `intro-types`: teaches about type annotations and variable declarations. Uses `const` to force learning of annotations, in an exception to the usual pattern.
- `intro-option`: teaches about basic usage of Option<T> and unwrapping to do more complex type conversions and working with uncertainty. The typical `unimplemented!("The exercise.")` stub should become familiar.

## Learning objectives

Students who complete these exercises should be able to...
- add comments for readability
- read [type signatures](https://en.wikipedia.org/wiki/Function_prototype)
- understand some consequences of type signatures, and thus
1. recognize some common primitive and a _few_ complex types
2. understand that type inference occurs
3. annotate types if inference fails
- declare variables using `let`
- declare variables using `const` (so as to force learning type annotation)
- perform some easier type casts
- use some conditional control flow, in particular to unwrap Option types, requiring:
1. handling `None`
2. extracting `value` from `Some(value)`
3. using `if` or `match` to do something based on that
- use `use` to access namespaces
- annotate or modify code when `rustc` asks for it
- write idiomatic functions
1. use implicit return
2. use semicolons as separators instead of terminators

That's a lot, but that's why it's multiple exercises.

## Out of scope

Student **should not** need to _completely_ understand, but _if_ it is pedagogically expedient, **may** discover mentions of
- arithmetic
- the relationship of char, str, and String
- `std::collections`, methods, or slices
- `let mut` vs. immutability
- `if let` usage
- **basic** usage of `match`
- a predefined macro or two like `format!`

Student **must not** need to understand
- Traits beyond `From` / `Into` / `TryFrom` / `TryInto`
- `fn main()` or handling IO
- ownership and lifetimes beyond `&str`
- More advanced `match` patterns
- More advanced control flow like loops and iterators
- General complex types, e.g. Structs, Enums, Tuples
- Result<T,E> and `panic!`
- `#[attributes]`, macros, or other compiler directives
- non-`std` crates

## Concepts

- `types`
- `fn`
- `option`
- `casting`
- `variables`
- `use`

## Prerequisites

None, hopefully.

## Resources to refer to

You can reference [The Book] as usual but in particular it is a good idea to read [Rust By Example] and [Rustlings].

### Hints

* Appropriate chapters of [The Book].
* The [standard library](https://doc.rust-lang.org/std/)

### After

* [Rust By Example]
* [Rustlings]

## Representer

No specific changes are expected to be required.

## Analyzer

No specific changes are expected to be required.

## Implementing

See our [implementation guide](https://github.com/exercism/v3/blob/master/languages/rust/howto/implementing-a-concept-exercise.md#how-to-implement-a-rust-concept-exercise) for guidance.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[The Book]: https://doc.rust-lang.org/book/
[Rust By Example]: https://doc.rust-lang.org/rust-by-example
[Rustlings]: https://github.com/rust-lang/rustlings

**Comment by efx** [comment]
@workingjubilee 👋 I'm dusting off some of the v3 issues and was curious if you still wanted to pursue this one? 

No problem if you do not. I'll simply remove the assignee and add the "help wanted" label for any potential future contributors.

**Comment by senekor** [comment]
After several failed attempts to create a high-quality syllabus, I believe any future attempts will have to do their own design work from scratch.
